internal/levee.go

```diff
@@ -39,75 +41,210 @@ func run(pass *analysis.Pass) (interface{}, error) {
+	for _, p := range fn.FreeVars {
+		switch t := p.Type().(type) {
+		case *types.Pointer:
+			if s, ok := utils.Dereference(t).(*types.Named); ok && conf.IsSource(s) {
+				graph.addSource(p.Name(), p)
 			}
+		}
+	}
 
-			for _, instr := range b.Instrs {
-				v, ok := instr.(*ssa.Call)
-				if !ok {
-					continue
+	for _, b := range fn.Blocks {
+		for _, inst := range b.Instrs {
+			value, _ := inst.(ssa.Value)
+
+			switch t := inst.(type) {
+			case *ssa.Alloc:
+				if conf.IsSource(utils.Dereference(t.Type())) && !source.IsProducedBySanitizer(t, conf) {
+					graph.addSource(value.Name(), t)
+				}
+
+			case *ssa.Call:
+				parameters := t.Call.Args
+				var parameterNames []string
+				for _, p := range parameters {
+					parameterNames = append(parameterNames, p.Name())
 				}
 				switch {
-				case conf.IsPropagator(v):
-					// Handling the case where sources are propagated to io.Writer
-					// (ex. proto.MarshalText(&buf, c)
-					// In such cases, "buf" becomes a source, and not the return value of the propagator.
-					// TODO Do not hard-code logging sinks usecase
-					// TODO  Handle case of os.Stdout and os.Stderr.
-					// TODO  Do not hard-code the position of the argument, instead declaratively
-					//  specify the position of the propagated source.
-					// TODO  Consider turning propagators that take io.Writer into sinks.
-					if a := getArgumentPropagator(conf, v); a != nil {
-						sources = append(sources, source.New(a, conf))
-					} else {
-						sources = append(sources, source.New(v, conf))
-					}
-
-				case conf.IsSink(v):
-					// TODO Only variadic sink arguments are currently detected.
-					if sinkVarargs := varargs.New(v); sinkVarargs != nil {
-						for _, s := range sources {
-							if sinkVarargs.ReferredBy(s) && !s.IsSanitizedAt(v) {
-								report(pass, s, v)
-								break
-							}
-						}
-					}
+				case conf.IsSink(t):
+					graph.addSink(t, parameterNames)
+				case conf.IsPropagator(t):
+					graph.addSource(value.Name(), t)
+				case conf.IsSanitizer(t):
+					graph.addSanitizer(t)
+				default:
+					graph.addCallEdges(*t, parameterNames)
 				}
+
+			case *ssa.FieldAddr:
+				if conf.IsSourceFieldAddr(t) {
+					graph.addSource(value.Name(), t)
+				}
+
+			case *ssa.IndexAddr:
+				x, _ := t.X.(ssa.Value)
+				// example instruction: t1 = &t0[0:int]
+				// the edge needs to be inverted since we are storing
+				// a reference; in effect, t1 is now an alias for t0
+				graph.addEdge(value.Name(), x.Name())
+
+			case *ssa.MakeInterface:
+				x := t.X.(ssa.Value)
+				graph.addEdge(x.Name(), value.Name())
+
+			case *ssa.Phi:
+				for _, e := range t.Edges {
+					graph.addEdge(e.Name(), value.Name())
+				}
+
+			case *ssa.Slice:
+				x, _ := t.X.(ssa.Value)
+				graph.addEdge(x.Name(), value.Name())
+
+			case *ssa.Store:
+				graph.addEdge(t.Val.Name(), t.Addr.Name())
+
+			case *ssa.UnOp:
+				x, _ := t.X.(ssa.Value)
+				graph.addEdge(x.Name(), value.Name())
```

Comment:
If I'm not mistaken, this addition is what now enables correct taint propagation to collections?  It feels as though this could have been its own PR.